graphbolt/src/cnumpy.cc

```diff
@@ -0,0 +1,217 @@
+  if (feature_fd_ == -1) {
+    throw std::runtime_error("npy_load: Unable to open file " + filename);
+  }
+
+  // Get system max thread number.
+  num_thd_ = torch::get_num_threads();
+  ring_ = new io_uring[num_thd_];
+
+  // Init io_uring queue.
+  for (int64_t t = 0; t < num_thd_; t++) {
+    io_uring_queue_init(group_size_, &ring_[t], 0);
+  }
+#endif
+}
+
+OnDiskNpyArray::~OnDiskNpyArray() {
+#ifdef __linux__
+  // IO queue exit.
+  for (int64_t t = 0; t < num_thd_; t++) {
+    io_uring_queue_exit(&ring_[t]);
+  }
+  close(feature_fd_);
+#endif
+}
+
+void OnDiskNpyArray::ParseNumpyHeader() {
+  FILE *fp = fopen(filename_.c_str(), "rb");
+  if (!fp)
+    throw std::runtime_error("npy_load: Unable to open file " + filename_);
+  char buffer[256];
+  size_t res = fread(buffer, sizeof(char), 11, fp);
+  if (res != 11) throw std::runtime_error("parse_npy_header: failed fread");
+  std::string header = fgets(buffer, 256, fp);
+  assert(header[header.size() - 1] == '\n');
+  prefix_len_ = 11 + header.size();
+
+  size_t loc1, loc2;
+
+  // Get shape location.
+  loc1 = header.find("(");
+  loc2 = header.find(")");
+  if (loc1 == std::string::npos || loc2 == std::string::npos)
+    throw std::runtime_error(
+        "parse_npy_header: failed to find header keyword: '(' or ')'");
+
+  std::regex num_regex("[0-9][0-9]*");
+  std::smatch sm;
+  std::string str_shape = header.substr(loc1 + 1, loc2 - loc1 - 1);
+  while (std::regex_search(str_shape, sm, num_regex)) {
+    feat_dim_.emplace_back(std::stoi(sm[0].str()));
+    str_shape = sm.suffix().str();
+  }
+
+  loc1 = header.find("descr");
+  if (loc1 == std::string::npos)
+    throw std::runtime_error(
+        "parse_npy_header: failed to find header keyword: 'descr'");
+  loc1 += 9;
+  bool littleEndian =
+      (header[loc1] == '<' || header[loc1] == '|' ? true : false);
+  assert(littleEndian);
+
+  std::string str_ws = header.substr(loc1 + 2);
+  loc2 = str_ws.find("'");
+  size_t word_size = atoi(str_ws.substr(0, loc2).c_str());
+
+  signed long feat_len = 1;
+  for (size_t i = 1; i < feat_dim_.size(); i++) {
+    feat_len *= feat_dim_[i];
+  }
+
+  feature_size_ = feat_len * word_size;
+}
+
+torch::Tensor OnDiskNpyArray::IndexSelectIOuring(torch::Tensor idx) {
+#ifdef __linux__
+  idx = idx.to(torch::kLong);
+  // The minimum page size to contain one feature.
+  int64_t align_len = (feature_size_ + ALIGNMENT) & (long)~(ALIGNMENT - 1);
+  int64_t num_idx = idx.numel();
+
+  char *read_buffer = (char *)aligned_alloc(
+      ALIGNMENT, (align_len + ALIGNMENT) * group_size_ * num_thd_);
+  char *result_buffer =
+      (char *)aligned_alloc(ALIGNMENT, feature_size_ * num_idx);
+
+  auto idx_data = idx.data_ptr<int64_t>();
+
+  // Record the inside offsets of feteched features.
+  int64_t residual[group_size_ * num_thd_];
+
+  // Indicator for index error.
+  bool error_flag = false;
+  torch::parallel_for(0, num_idx, group_size_, [&](int64_t begin, int64_t end) {
```

Comment:
Before we call any parallel_for, we need to add a check so that the number of threads is the same as in the constructor, such as:
`TORCH_CHECK(num_thd_ == torch::get_num_threads(), "The number of threads can not be changed after a disk feature fetcher is constructed.");`